### ISSUE TITLE
Feature request: auto config for dynamic apis

### ISSUE THREAD

**Issue by richardeschloss** [richardeschloss/nuxt-socket-io#74]
It may be possible to make apis completely dynamic and have the emitter configuration be automated based on the methods the IO client and server expose. Currently, the namespace config is entirely manual, but can possibly be automated if we think of each side like this:

server.js:
```
const svc = Object.freeze({
  method1(args) { ... },
  method2(ags) { ... }
})

const serverApi = Object.keys(svc)
socket.emit('serverApi', serverApi) // sends client its api
```

Similarly, the client may have methods it exposes and events it listens for. It may wish to communicate this to the server.

**Comment by richardeschloss** [comment]
Just taking a note: one person on Medium expressed an interest in perhaps a support for "vuex actions/mutation prefixes", similar to vuex-socket-io, but I'm not convinced on that design approach just yet. (I'll think about it)

**Comment by richardeschloss** [comment]
Update: I am working on a local branch...this is more challenging than I thought it would be, but I think I'm making progress.

**Comment by richardeschloss** [comment]
Got some cool shit pushed to [feat/dynamicApi](https://github.com/richardeschloss/nuxt-socket-io/tree/feat/dynamicApi) but not quite sure what word to emphasize more at the moment: "cool" or "shit" :). I think it'll become more cool in the next few days and less shit....fingers crossed.

**Comment by richardeschloss** [comment]
Status update: server methods auto-gen feature works, with cache control option.

**Comment by richardeschloss** [comment]
Status update: added code to dynamically register server events (with option to ignore)

**Comment by richardeschloss** [comment]
Status update: 
* had to rearrange the api data structure a little bit, so that the client knows where to store event data cleanly. 
* so far, some really cool stuff working at the page/component level, but next phase of this will be to somehow integrate this into Vuex too (with options). So, fortunately, vuex's api lets me dynamically register module(s), so I may consider creating a "$nuxtSocket" vuex module that would contain the socket instance and apis. The api tree(s) can get pretty deep, but I'm trying to implement this as a clean solution for users. 

If there's an audience reading this, I apologize for taking so long to merge this...the branch feat/dynamicApi contains some most of the code, even though the docs aren't written yet, I think the code has been made somewhat readable / understandable (I hope).

**Comment by richardeschloss** [comment]
Status update: ok, things are really cool. The code is just messy and needs to be cleaned up and properly tested. (Plus as I'm going through this, I'm jotting down implementation ideas for the other features to improve flexibility).  The PR might be larger than expected, but worth it.

**Comment by richardeschloss** [comment]
Status update. Feature is pretty much done and it's fucking awesome (I think). My absolute favorite feature of the entire project. I just gotta update the tests for coverage (a timely process since this is a big feature), and then write docs to explain how to use it. I think communicating that will be the biggest challenge actully, but once people understand it, I think they'll like this feature.

**Comment by richardeschloss** [comment]
Merged in #87 . Updated docs and articles to come (a lot to explain, and I hope I can communicate it so people understand it)
